README.md
![EVEApi for ruby](https://github.com/aladac/eveapi/raw/master/doc/eveapi.png)

[![Gem Version](https://badge.fury.io/rb/eveapi.svg)](http://badge.fury.io/rb/eveapi)
[![Build Status](https://secure.travis-ci.org/aladac/eveapi.svg?branch=master)](https://travis-ci.org/aladac/eveapi)
[![Code Climate](https://codeclimate.com/github/aladac/eveapi/badges/gpa.svg)](https://codeclimate.com/github/aladac/eveapi)
[![Test Coverage](https://codeclimate.com/github/aladac/eveapi/badges/coverage.svg)](https://codeclimate.com/github/aladac/eveapi/coverage)
[![Downloads](https://img.shields.io/gem/dt/eveapi.svg)](https://rubygems.org/gems/eveapi)
[![Latest Version Downloads](https://img.shields.io/gem/dtv/eveapi.svg)](https://rubygems.org/gems/eveapi)

## TOC
- [Description](#desc)
- [Documentation](#doc)
- [Install](#install)
- [Working methods](#methods)
  - [`Client`](#client)
  - [`Character`](#character)
  - [`Alliance`](#alliance)
  - [`Crest`](#crest_methods)
- [CREST](#crest)
- [Features](#feats)
- [Examples](#examples)
- [Requirements](#reqs)
- [Contributing](https://github.com/aladac/eveapi/blob/master/CONTRIBUTING.md)

<a name="name"></a>
## Description
EVE Online API Client for ruby

<a name="doc"></a>
## Documentation
- [rubydoc.info](http://rdoc.info/gems/eveapi)
- [legacy methods (wiki)](https://github.com/aladac/eveapi/wiki)

<a name="install"></a>
## Install
    $ gem install eveapi

<a name="methods"></a>
## Working methods
Working `Client` methods names can be listed by calling `EVEApi::Client#working_methods`.

Check out [rubydoc.info](http://rdoc.info/gems/eveapi) for the documentation.

*Legacy methods are available [here](https://github.com/aladac/eveapi/wiki)*

<a name="client"></a>
#### Client

- [`EVEApi::Client.new`](http://www.rubydoc.info/gems/eveapi/EVEApi/Client)
- [`EVEApi::Client#characters`](http://www.rubydoc.info/gems/eveapi/EVEApi/Client#characters-instance_method)
- [`EVEApi::Client#server_status`](http://www.rubydoc.info/gems/eveapi/EVEApi/Client#server_status-instance_method)
- [`EVEApi::Client#call_list`](http://www.rubydoc.info/gems/eveapi/EVEApi/Client#call_list-instance_method)
- [`EVEApi::Client#working_methods`](http://www.rubydoc.info/gems/eveapi/EVEApi/Client#working_methods-instance_method)
- [`EVEApi::Client#account_status`](http://www.rubydoc.info/gems/eveapi/EVEApi/Client#account_status-instance_method)
- [`EVEApi::Client#key_info`](http://www.rubydoc.info/gems/eveapi/EVEApi/Client#key_info-instance_method)

<a name="characters"></a>
#### Character

- [`EVEApi::Character#wallet_journal`](http://www.rubydoc.info/gems/eveapi/EVEApi/Character#wallet_journal-instance_method)
- [`EVEApi::Character#contracts`](http://www.rubydoc.info/gems/eveapi/EVEApi/Character#contracts-instance_method)
- [`EVEApi::Character#wallet_transactions`](http://www.rubydoc.info/gems/eveapi/EVEApi/Character#wallet_transactions-instance_method)
- [`EVEApi::Character#upcoming_calendar_events`](http://www.rubydoc.info/gems/eveapi/EVEApi/Character#upcoming_calendar_events-instance_method)
- [`EVEApi::Character#standings`](http://www.rubydoc.info/gems/eveapi/EVEApi/Character#standings-instance_method)
- [`EVEApi::Character#skill_queue`](http://www.rubydoc.info/gems/eveapi/EVEApi/Character#wallet_journal-instance_method)
- [`EVEApi::Character#skill_in_training`](http://www.rubydoc.info/gems/eveapi/EVEApi/Character#skill_in_training-instance_method)
- [`EVEApi::Character#research`](http://www.rubydoc.info/gems/eveapi/EVEApi/Character#research-instance_method)
- [`EVEApi::Character#notifications`](http://www.rubydoc.info/gems/eveapi/EVEApi/Character#notifications-instance_method)
- [`EVEApi::Character#medals`](http://www.rubydoc.info/gems/eveapi/EVEApi/Character#medals-instance_method)
- [`EVEApi::Character#market_orders`](http://www.rubydoc.info/gems/eveapi/EVEApi/Character#market_orders-instance_method)
- [`EVEApi::Character#mail_messages`](http://www.rubydoc.info/gems/eveapi/EVEApi/Character#mail_messages-instance_method)
- [`EVEApi::Character#mailing_lists`](http://www.rubydoc.info/gems/eveapi/EVEApi/Character#mailing_lists-instance_method)
- [`EVEApi::Character#industry_jobs`](http://www.rubydoc.info/gems/eveapi/EVEApi/Character#industry_jobs-instance_method)
- [`EVEApi::Character#contact_notifications`](http://www.rubydoc.info/gems/eveapi/EVEApi/Character#contact_notifications-instance_method)
- [`EVEApi::Character#contact_list`](http://www.rubydoc.info/gems/eveapi/EVEApi/Character#contact_list-instance_method)
- [`EVEApi::Character#character_sheet`](http://www.rubydoc.info/gems/eveapi/EVEApi/Character#character_sheet-instance_method)
- [`EVEApi::Character#asset_list`](http://www.rubydoc.info/gems/eveapi/EVEApi/Character#asset_list-instance_method)
- [`EVEApi::Character#account_balance`](http://www.rubydoc.info/gems/eveapi/EVEApi/Character#account_balance-instance_method)

<a name="crest_methods"></a>
#### Crest

- [`EVEApi::Crest.new`](http://www.rubydoc.info/gems/eveapi/EVEApi/Crest)
- [`EVEApi::Crest#alliances`](http://www.rubydoc.info/gems/eveapi/EVEApi/Crest#alliances-instance_method)
- [`EVEApi::Crest#types`](http://www.rubydoc.info/gems/eveapi/EVEApi/Crest#types-instance_method)

<a name="alliance"></a>
#### Alliance

- [`EVEApi::Alliance.find(id = nil)`](http://www.rubydoc.info/gems/eveapi/EVEApi/Alliance#find-instance_method)
- [`EVEApi::Alliance#info`](http://www.rubydoc.info/gems/eveapi/EVEApi/Alliance#info-instance_method)
- [`EVEApi::Alliance#corporations`](http://www.rubydoc.info/gems/eveapi/EVEApi/Alliance#corporations-instance_method)
- [`EVEApi::Alliance#to_h`](http://www.rubydoc.info/gems/eveapi/EVEApi/Alliance#to_h-instance_method)

<a name="crest"></a>
## CREST

Rudimentary CREST support is added via the `EVEApi::Crest` class.
Working mehtods


```ruby
a = Crest.new.alliances.select { |a| a.name == "Goonswarm Federation" }.first
=> #<EVEApi::Alliance:0x007fd511c48248 ...>
a.to_h
=> {
          :href => "https://public-crest.eveonline.com/alliances/1354830081/",
        :id_str => "1354830081",
    :short_name => "CONDI",
          :name => "Goonswarm Federation",
            :id => 1354830081
}
a.info['url']
=> "http://www.goonfleet.com"
a.info.keys
=> [
    [ 0] "startDate",
    [ 1] "corporationsCount",
    [ 2] "description",
    [ 3] "executorCorporation",
    [ 4] "primeHour_str",
    [ 5] "primeHour",
    [ 6] "deleted",
    [ 7] "corporationsCount_str",
    [ 8] "creatorCorporation",
    [ 9] "url",
    [10] "id_str",
    [11] "creatorCharacter",
    [12] "capitalSystem",
    [13] "corporations",
    [14] "shortName",
    [15] "id",
    [16] "name"
]
```

<a name="feats"></a>
## Features

Uses [excon](https://github.com/excon/excon),  [crack](https://github.com/jnunemaker/crack), `method_missing` and a couple of rescue blocks to automate access to EVE Online API.
- returns results as `Hash` or `Array`
- raises exceptions with messages from the API itself
- converts ruby methods like `account_api_info` to an EVE Online API request like `GET /account/APIKeyInfo.xml.aspx` along with query params.

<a name="examples"></a>
## Examples

### Getting Implant names for character
```ruby
require 'eveapi'
# initialize the client with key_id, vcode, character_id
client = Client.new(4278167, "supersecretstuff", 95512059)
# => #<EVEApi::Client:0x007fe349ad0d48 ... >
client.characters.first.character_sheet[:implants].map { |i| i[:type_name] }
# => [
#     [0] "Social Adaptation Chip - Basic",
#     [1] "Cybernetic Subprocessor - Standard",
#     [2] "Memory Augmentation - Standard",
#     [3] "Neural Boost - Standard",
#     [4] "Ocular Filter - Standard"
# ]
```

### Server status
```ruby
require 'eveapi'
client.server_status
# => {
#        :server_open => "True",
#     :online_players => "32500"
# }
```

### Showing accounts characters

```ruby
require 'eveapi'
client = EVEApi::Client.new
client.key_id = 4278167
client.vcode = "7QJg6p5BZNpDBp2FIz39dGwa7jnNaXAuYyLUVitlTQ3rY60VPBcaTpJVfYIkiW5l"
client.characters
# => {
#                 :name => "Quint Slade",
#         :character_id => "95512059",
#     :corporation_name => "Imperial Academy",
#       :corporation_id => "1000166",
#          :alliance_id => "0",
#        :alliance_name => "",
#           :faction_id => "0",
#         :faction_name => ""
# }
```

## Requirements
<a name="reqs"></a>
`excon` and `crack` automatically installed as dependencies during `gem install`, or `bundle install` in development.

## Copyright

Copyright (c) 2015 Adam Ladachowski

See [LICENSE.txt](https://github.com/aladac/eveapi/blob/master/LICENSE.txt) for details.


ChangeLog.md
### 0.1.0 / 2015-09-09

* Initial release:
Several methods are fully working. Need to look into the ones that do not.
Complete list of working methods in the README.md

### 0.1.1 / 2015-09-10
* Rudimentary CREST support
Added a Crest class to handle CREST access
- alliances method is the only one available
- alliances returns an Array of aliances


lib/eveapi/client.rb
module EVEApi
  # Client instance and HTTP method handling
  class Client
    # HTTP connection to the API endpoint
    # @return [Excon::Connection]
    attr_accessor :connection
    # API Key ID
    # @return [String]
    attr_accessor :key_id
    # API Key verification code
    # @return [String]
    attr_accessor :vcode
    # Character ID
    # @return [String]
    attr_accessor :character_id
    # Number of result rows to get
    # @return [String]
    attr_accessor :row_count

    # @param [String] key_id API key ID
    # @param [String] vcode API verification code
    # @param [String] character_id Character ID
    # @return [EVEAPI::Client] client
    def initialize(key_id = nil, vcode = nil, character_id = nil)
      @connection ||= Excon.new(API_ENDPOINT)
      @key_id = key_id
      @character_id = character_id
      @vcode = vcode
    end

    # Query params used in the API request
    #
    # @return [Hash] query params for the API request
    def params
      {
        'rowCount'    => row_count,
        'keyID'       => key_id,
        'vCode'       => vcode,
        'characterID' => character_id
      }.select { |_k, v| v }
    end

    # API methods
    #
    # @return [Array] list of ruby API method names
    # @example
    #   client.api_methods
    #   # => [
    #   # [ 0] :char_chat_channels,
    #   # [ 1] :char_bookmarks,
    #   # [ 2] :char_locations,
    #   # [ 3] :char_contracts,
    #   # ...
    #   # ]
    def api_methods
      api_methods_hash.map { |m| m[:name] }
    end

    def api_methods_hash
      api_call_list[:calls].map do |m|
        { name: m.ruby_method_name, desc: m[:description] }
      end
    end
    private :api_methods_hash

    def characters_array(account_characters)
      account_characters.to_a.map do |character|
        args = character
        args.merge!(key_id: key_id, vcode: vcode)
        Character.new(args)
      end
    end
    private :characters_array

    def characters_hash(account_characters)
      args = account_characters
      args.merge!(key_id: key_id, vcode: vcode)
      [Character.new(args)]
    end
    private :characters_hash

    # List of characters belonging to the account
    #
    # @return [Array] of characters for account
    # @see Character
    # @example
    #   client = Client.new(4278167, "supersecretstuff", 95512059)
    #   characters = client.characters
    #   characters.first.name
    #   # => "Quint Slade"
    def characters
      case account_characters
      when Array
        return characters_array(account_characters)
      when Hash
        return characters_hash(account_characters)
      end
    end

    # Status of the Tranquility server
    #
    # @return [Hash] Tranquility server status
    # @example
    #   Client.new.server_status
    #   # => {
    #   #        :server_open => "True",
    #   #     :online_players => "25292"
    #   # }
    def server_status
      api_request(:server_server_status)
    end

    # Account Status
    #
    # @return [Hash] account status
    # @example
    #   client.account_status
    #   # => {
    #   #                   :paid_until => "2015-10-12 23:55:48",
    #   #                  :create_date => "2006-09-10 02:17:00",
    #   #                  :logon_count => "2276",
    #   #                :logon_minutes => "193996",
    #   #     :multi_character_training => nil,
    #   #                       :offers => nil
    #   # }
    def account_status
      api_request(:account_account_status)
    end

    # API key details
    #
    # @return [Hash] API key info
    # @example
    #    client = Client.new(4278167, "supersecretstuff", 95512059)
    #    client.key_info
    #    # => {
    #    #     :access_mask => "268435455",
    #    #            :type => "Account",
    #    #         :expires => "",
    #    #      :characters => {
    #    #             :character_id => "95512059",
    #    #           :character_name => "Quint Slade",
    #    #           :corporation_id => "1000166",
    #    #         :corporation_name => "Imperial Academy",
    #    #              :alliance_id => "0",
    #    #            :alliance_name => "",
    #    #               :faction_id => "0",
    #    #             :faction_name => ""
    #    #     }
    #    # }
    def key_info
      api_request(:account_api_key_info)
    end

    # Description of method
    # Details
    # @return [Hash] list of API +:calls+ and +:call_groups+
    # @example
    #   Client.new.call_list.keys
    #   # => [
    #   #     [0] :call_groups,
    #   #     [1] :calls
    #   #    ]
    #   Client.new.call_list[:call_groups]
    #   # => [
    #   # [0] {
    #   #    :group_id => "1",
    #   #        :name => "Account and Market",
    #   # :description => "Market Orders, account balance and journal history."
    #   # },
    #   # ... ]
    def call_list
      api_request(:api_call_list)
    end

    def api_request(name)
      http = connection.get(path: name.to_path, query: params)
      request = EVEApi::Request.new(http)
      request.result
    end
    private :api_request

    # Dynamic handling of API requests
    def method_missing(name, *_args, &_block)
      fail 'Invalid Method Name' if name.to_path.empty?
      api_request(name)
    end
    private :method_missing

    # List of implemented methods
    #
    # @return [Array] list of implemented +Client+ methods
    # @example
    #   Client.new.working_methods
    #   # => [
    #   #     [0] :account_status,
    #   #     [1] :server_status,
    #   #     [2] :characters,
    #   #     [3] :api_methods,
    #   #     [4] :key_info,
    #   #     [5] :call_list
    #   # ]
    def working_methods
      EVEApi::WORKING_METHODS
    end
  end
end
